### PR TITLE
[cmake] Add support for partially enabled modules

### LINKED ISSUE #12591: Allow partial enablement of modules
Commit 46ba2eefd00371b024ae239626c7600cd3259885 (https://github.com/root-project/root/pull/12082) enabled a few modules for Windows, but added `NO_CXXMODULE` to a couple others that don't work yet. This requires support for figuring out which dependencies have modules and which ones don't. Unfortunately, this support had some flaws and had to be reverted in commit 2d7373d47ef2acc53f4eed004082fee83caab6a3 (https://github.com/root-project/root/pull/12561) - see the summary and commit messages for more details.

To summarize, the core of the problem is that the build system expected a dependency to have a module if it had been registered beforehand. This is incompatible with how we build `libCore` and some other fundamental libraries. We likely have to implement it the other way around, ie keep track of which targets will produce a module. This may be enough and work for the use cases we have on Windows.

### PR DESCRIPTION
**Pull request by hahnjo**
On Windows, we currently cannot generate all modules due to crashes, so some dictionaries are generated with the option `NO_CXXMODULE`. This information needs to be propagated and taken into account when those dictionaries are, in turn, specified as dependencies.

This re-implements the logic reverted in commit 2d7373d47e (see the message for details) and closes #12591.

### CODE REVIEW

**Comment by hahnjo** [comment]
@bellenot @vgvassilev could you check if this change works with C++ modules on Windows? It's the easiest approach that I could come up, but maybe that already works...

**Comment by bellenot** [comment]
So FYI, the C++ modules on Windows don't work at all with the latest version of Visual Studio (like always when updating it), it cannot even generate `Core.pcm`

**Comment by vgvassilev** [comment]
> So FYI, the C++ modules on Windows don't work at all with the latest version of Visual Studio (like always when updating it), it cannot even generate `Core.pcm`

What is the current issue?

**Comment by bellenot** [comment]
> What is the current issue?

```
Generating G__Core.cxx, ../bin/Core.pcm
While building module 'Core':
While building module 'std' imported from input_line_1:1:
In file included from <module-includes>:1:
In file included from C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\algorithm:11:
In file included from C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\xmemory:14:9>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\new(112,55): error GAE4BF799: declaration of 'set_new_handler' has a different language linkage [C:\Users\bellenot\build\x64\cxxmodules\core\G__Core.vcxproj]
_EXPORT_STD extern "C++" _CRTIMP2 new_handler __cdecl set_new_handler(_In_opt_ new_handler) noexcept;
                                                      ^
C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\ucrt\new.h:32:42: note: previous declaration is here
            _CRTIMP2 new_handler __cdecl set_new_handler(_In_opt_ new_handler _NewHandler) throw();
                                         ^
While building module 'Core':
While building module 'std' imported from input_line_1:1:
In file included from <module-includes>:66:
In file included from C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\random:12:9>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\__msvc_int128.hpp(115,20): error G4BA35A91: call to '_addcarry_u64' is ambiguous [C:\Users\bellenot\build\x64\cxxmodules\core\G__Core.vcxproj]
            return _addcarry_u64(_Carry, _Left, _Right, &_Result);
                   ^~~~~~~~~~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\intrin0.inl.h:285:36: note: candidate function
__MACHINEX64(unsigned char __cdecl _addcarry_u64(unsigned char, unsigned __int64, unsigned __int64, unsigned __int64 *))
                                   ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\intrin0.inl.h:43:32: note: expanded from macro '__MACHINEX64'
#define __MACHINEX64           __MACHINE
                               ^
C:/Users/bellenot/build/x64/cxxmodules/etc/cling\lib\clang\13.0.0\include\adxintrin.h:47:1: note: candidate function
_addcarry_u64(unsigned char __cf, unsigned long long __x,
^
While building module 'Core':
While building module 'std' imported from input_line_1:1:
In file included from <module-includes>:66:
In file included from C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\random:12:9>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\__msvc_int128.hpp(129,20): error G4BA35A91: call to '_subborrow_u64' is ambiguous [C:\Users\bellenot\build\x64\cxxmodules\core\G__Core.vcxproj]
            return _subborrow_u64(_Carry, _Left, _Right, &_Result);
                   ^~~~~~~~~~~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\intrin0.inl.h:286:36: note: candidate function
__MACHINEX64(unsigned char __cdecl _subborrow_u64(unsigned char, unsigned __int64, unsigned __int64, unsigned __int64 *))
                                   ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\intrin0.inl.h:43:32: note: expanded from macro '__MACHINEX64'
#define __MACHINEX64           __MACHINE
                               ^
C:/Users/bellenot/build/x64/cxxmodules/etc/cling\lib\clang\13.0.0\include\adxintrin.h:63:1: note: candidate function
_subborrow_u64(unsigned char __cf, unsigned long long __x,
^
input_line_1:1:10: fatal error: could not build module 'std'
#include <new>
 ~~~~~~~~^
Warning in cling::IncrementalParser::CheckABICompatibility():
  Failed to extract C++ standard library version.
Warning in cling::IncrementalParser::CheckABICompatibility():
  Possible C++ standard library mismatch, compiled with _CRT_MSVCP_CURRENT 'msvcp140.dll'
  Extraction of runtime standard library version was: ''
While building module 'Core':
While building module 'Cling_Runtime' imported from input_line_2:1:
While building module 'Cling_Runtime_Extra' imported from C:/Users/bellenot/build/x64/cxxmodules/etc\cling/Interpreter/RuntimeUniverse.h:27:
In file included from <module-includes>:1:
C:/Users/bellenot/build/x64/cxxmodules/etc\cling/Interpreter/DynamicExprInfo.h:13:10: fatal error: could not build module 'std'
#include <string>
 ~~~~~~~~^
C:/Users/bellenot/build/x64/cxxmodules/etc\cling/Interpreter/RuntimeUniverse.h:27:10: fatal error: could not build module 'Cling_Runtime_Extra'
#include "cling/Interpreter/Visibility.h"
 ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
While building module 'Core':
While building module 'ROOT_Rtypes' imported from C:/Users/bellenot/build/x64/cxxmodules/include\Rtypes.h:23:
In file included from <module-includes>:1:
C:/Users/bellenot/build/x64/cxxmodules/include/RtypesCore.h:28:10: fatal error: could not build module 'std'
#include <cstddef> // size_t, NULL
 ~~~~~~~~^
In file included from input_line_3:2:
C:/Users/bellenot/build/x64/cxxmodules/include\Rtypes.h:23:10: fatal error: could not build module 'ROOT_Rtypes'
#include "RtypesCore.h"
 ~~~~~~~~^~~~~~~~~~~~~~
While building module 'Core':
While building module 'ROOT_Foundation_C' imported from C:/Users/bellenot/build/x64/cxxmodules/include\Rtypes.h:29:
In file included from <module-includes>:1:
C:/Users/bellenot/build/x64/cxxmodules/include/ThreadLocalStorage.h:60:10: fatal error: could not build module 'ROOT_Rtypes'
#include "RtypesCore.h"
 ~~~~~~~~^~~~~~~~~~~~~~
While building module 'Core':
While building module 'ROOT_Foundation_Stage1_NoRTTI' imported from C:/Users/bellenot/build/x64/cxxmodules/include/Rtypes.h:195:
In file included from <module-includes>:2:
In file included from C:/Users/bellenot/build/x64/cxxmodules/include/ROOT/RStringView.hxx:25:
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\string_view:9:10: fatal error: could not build module 'std'
#include <yvals.h>
 ~~~~~~~~^
```

**Comment by vgvassilev** [comment]
We probably need an entry for `xmemory` in the modulemap..

**Comment by hahnjo** [comment]
ping, what do we do with this PR?

**Comment by vgvassilev** [comment]
I still think this is the right way to go for two reasons. It enables windows but also improves the migration experience for external codebases. Of course it would have been better to fix the underlying modules issue on windows.

**Comment by hahnjo** [comment]
Ok, then let me rebase this PR and run it through CI...

**Comment by hahnjo** [comment]
ping @bellenot

**Review by bellenot — APPROVED** [review]
@hahnjo I let @vgvassilev take the final word on this PR, since the C++ modules don't work on Windows. But in principle, if it works everywhere else, it's fine with me.

**Comment by hahnjo** [comment]
To be clear, this would just be a first step towards (partially enabled) modules on Windows

File: cmake/modules/RootMacros.cmake [hunk 1]
```diff
@@ -568,7 +568,10 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
 
   # modules.idx deps
   get_property(local_modules_idx_deps GLOBAL PROPERTY modules_idx_deps_property)
+  get_property(local_no_cxxmodules GLOBAL PROPERTY no_cxxmodules_property)
   if (ARG_NO_CXXMODULE)
+    list(APPEND local_no_cxxmodules ${cpp_module})
```
[the remark this answers is not in the document]
Comment (pcanal) [reply cmake/modules/RootMacros.cmake:573]: Actually this might be okay.  The pattern that is a problem was replaced in commit 1ab0815203

**Pull request merged**